woocommerce/woocommerce
# Use Mozart instead of a script to renamespace packages

### All Submissions:


### Changes proposed in this Pull Request:

The dependency injection container [has been recently reinstated ](https://github.com/woocommerce/woocommerce/pull/27733) by using a provisional script that modifies the `league/container` package in-place (in the `vendor` directory) by using regular expressions.

This pull requests removes this custom script and uses instead [a custom fork of the coenjacobs/mozart package](https://github.com/woocommerce/mozart). The custom forks implements two extra features that are not present in the original package:

* [Add a mechanism to exclude packages from processing](https://github.com/coenjacobs/mozart/pull/76) - because we don't want to renamespace the `psr/container` package.
* [Delete empty vendor directories after deleting package directories](https://github.com/coenjacobs/mozart/pull/77) - so that empty directories aren't left hanging in `vendor`.

The package is loaded directly from the repository of the fork, targetting a branch that has the branches from these two pull requests merged in.

**Update:** The two pull requests were approved and merged into the original Mozart repository, so now the package is loaded directly from there, not from the fork.

The result is that now the `league/container` package is moved to the `src/Internal/Vendor` directory, with the appropriate PSR-4 namespace, by running `composer run move-vendor-packages`; the resulting moved package is added to source control, so it's not necessary to run the command again except for update to a newer version of the package after a `composer update`. Additional packages can be moved/renamespaced in the future if needed, by just adding their names to the `extra/mozart/packages` key in `composer.json`.

### How to test the changes in this Pull Request:

1. Start clean: delete the `vendor` directory and everything inside `src/Internal/Vendor` except the `README` file.
2. Run `composer install`. Verify that the `vendor` directory contains `league` and `psr` directories.
3. Run `composer run move-vendor-packages`. Verify that a new `src/Internal/Vendor/League` directory exists. Also `league`  will have disappeared from `vendor` (Mozart did this).
4. Smoketest (it's enoguh to verify that the site loads at all).
5. Run `composer install` again to bring back `vendor/league`. Run `composer install --no-dev` and verify that `vendor/league` has disappeared and the site still loads.
6. Run `npm run build` and verify that the generated zip doesn't have a `vendor/league` directory (but still has `vendor/psr`).
7. Smoketest the generated zip file.

### Other information:


### Changelog entry

> Dev - Use the coenjacobs/mozart package to renamespace vendor packages.
